### ISSUE TITLE
static service injection analysis

### ISSUE THREAD

**Issue by Windvis** [embroider-build/embroider#1079]
A while ago I opened https://github.com/embroider-build/embroider/issues/829 because I didn't understand why a service was not included in the build while it was explicitly injected in multiple locations. This brought up the discussion of a potential `staticServices` flag for Embroider, and creating an API for injecting services that is easier to statically analyze. It's been a while since I've closed that issue but I still think it would a good addition so I'm opening this one to track it.

**Comment by ef4** [comment]
I think it may be helpful to sketch how you could "manually" get services to be statically connected to where they're used, because it's also how you could implement a staticServices mode in Embroider.

Let's say you have an app building with Embroider and even using route splitting. You have `app/services/expensive` which is only used on a couple routes, and depends on a huge library, so you'd rather only load it when it's needed.

1. Make an `app/lib` directory and set the embroider option `staticAppPaths: ['lib']`. [See inline documentation](https://github.com/embroider-build/embroider/blob/47a1455ae83d690451abd827a6d6c44e30a28583/packages/core/src/options.ts#L38-L60). 
2. Move `app/services/expensive.js` to `app/lib/expensive.js`.
3. Now Ember definitely can't find the service and it won't be in your build.
4. In the files that want to use the service, do:
    ```js
    import Expensive from '../lib/expensive';
    window.define('my-app/services/expensive', function{ return Expensive; });
    export default class extends Component {
      @service expensive;
    }
    ```

This would give you just-in-time registration of the service.

The main work of implementing a staticServices mode in embroider is doing this translation automatically. When we see the injection in a file, we'd insert the `import` and `define` too. And under this mode, we would automatically treat `app/services` as if it was in `appStaticPaths`, so that we don't eagerly include any services.

**Comment by mydea** [comment]
Not sure if this is "nice", but we use this in our app to register some services only on-demand:

```js
// my-code-split-route.js
import MySplitService from 'my-app/split-services/my-split-service';

export default class MyCodeSplitRoute extends Route {
  constructor(owner) {
    super(owner);

    owner.register(
      'service:my-split-service',
      MySplitService
    );
  }
}
```

Then I can use the service normally everywhere. That's of course a more manual (and for sure a bit brittle) approach, but it works :sweat_smile: We use this for some special services that are only used in an `admin` subroute, and there we use a very well defined set of components etc. that make it feasible to handle this manually.

**Comment by tansongyang** [comment]
@ef4 Would that solution you proposed be compatible with service unit tests?

**Comment by ef4** [comment]
Yes, as long as the unit test also imports and registers the service, the same way the other consumers of the service would.

**Comment by tansongyang** [comment]
Oh, just thought of one more question @ef4. I wasn't sure based on your earlier comment, but is this actually on the Embroider roadmap?
